### ISSUE TITLE
Peer Review by John Torres jt072

### ISSUE THREAD

**Issue by john2408** [hempelt/project#1]
### Peer review team

- Peer review by: John Torres
- Names of team members that participated in this review: John Torres

### Goal of the project

The goal of this project is to analyze the impact of different experimental conditions on the measured viscosity of biopharmaceutical drug products. The project aims to identify the relationship between various predictors (such as temperature, product concentration, protein format and molecular weight) and the response variable (viscosity). Ultimately, the project seeks to develop a predictive model that can estimate the viscosity of new drug products without the need for extensive laboratory experiments, thereby reducing development time and costs.

### Data

- Describe the data used or collected, if any. Is the data adequate for the project?

The data used in this project consists of viscosity measurements for various biopharmaceutical products. Each observation in the dataset corresponds to one measurement value. The dataset includes the following variables:

- **viscosity_mPas**: Measured viscosity of the sample in mPas (response variable).
- **replicate**: Number of replicates for each measurement.
- **entered_on**: Date when the measurement was conducted.
- **instrument:** Instrument used to measure the viscosity.
- **temperature_c**: Temperature at which the measurement was conducted.
- **product_concentration_mg_mL**: Concentration of the product in the aqueous solution in mg/mL.
- **product**: Internal product name as a unique code.
- **protein_format**: Protein format of the investigated product.
- **molecular_weight_kda**: Molecular weight of the investigated product in kDa.
- **extinction_coefficient_l_molcm**: Extinction coefficient of the investigated product in L·mol⁻¹·cm⁻¹.
- **isoelectric_point**: Isoelectric point of the investigated product.

The data appears to be adequate for the project as it includes multiple predictors that are likely to impact the response variable (viscosity). The dataset covers different experimental conditions such as temperature and product concentration, which are essential for building a predictive model.

### Approach, tools and methods

The project will use the following approaches, tools, and methods:

**Approaches**:
1. **Data Exploration and Visualization**: Initial exploration of the dataset to understand the distribution and relationships between variables.
2. **Predictive Modeling:** Building a model to predict the viscosity of new drug products based on the predictors.

**Tools**:

Python Libraries:
- pandas: For data manipulation and analysis.
- numpy: For numerical operations.
- altair:  For data visualization.
- scikit-learn: For machine learning and predictive modeling.
- pickle: For saving and loading models.
- datetime

Methods:
Data Preprocessing: Cleaning and preparing the data for analysis.
Train-Test Split: Splitting the data into training and testing sets using `train_test_split` from scikit-learn.
Linear Regression: Using `LinearRegression` from scikit-learn to build a predictive model.
Cross-Validation: Evaluating the model's performance using ´cross_val_score´ from scikit-learn.

These approaches, tools, and methods will help in analyzing the data, understanding the relationships between variables, and building a robust predictive model for viscosity.

### Lack of clarity

- Is there anything that is unclear from the proposal? 

The draft analysis is generally clear, but there are a few points that could benefit from further clarification like additional data preprocessing steps.

### Possible improvements

- Provide constructive feedback on how the team might be able to improve their project.Make sure your feedback includes at least one comment on the modeling aspect of the project, but do feel free to comment on aspects beyond the modeling.

- **Data Preprocessing Steps**:  It would be helpful to outline how missing values and outliers would be handled. This could improve the overall model accuracy.
- **Categorical Features**: Include categorical features as dummy variables in the model. For Example: `instrument` and `protein_format`.
- **Model Selection**: The model accuracy could be improved by applying a `backward elimination` of `forward selection` of features as explained in the Chaper 8.4 of the book [Introduction to Modern Statistics](https://openintro-ims.netlify.app/model-mlr#sec-model-selection).
- ** Feature Engineering Interaction Terms**: Consider creating interaction terms between predictors (e.g., temperature and concentration) to capture more complex relationships.
- **Normalization/Standardization**: Apply normalization or standardization to numerical to ensure they are on a similar scale, which can improve model performance. This is done before training the model. 

### Presentation

- What aspect of this project are you most interested in and would like to see highlighted in the final presentation?

I would like the project to explore the influence of the categorical predictors in the model accuracy. 

If we analyze the varialbes `instrument` and `protein_format`, using the function `.count_values()` we can identify the following distributions: 

`instrument`
VISCOSIMETER_02    324
VISCOSIMETER_01    178
Name: count, dtype: int64

`protein_format`
IgG2         250
DoppelMab    108
Knob/Hole     72
IgG4          72

Both contain a good distribution of data, which might be helpful to improve the model's accuracy. These variables can be transformed to numerical variables using the function `pd.get_dummies`, for example:

```python
df = pd.get_dummies(df, columns=['instrument'])
```
Additional, exploring different models using the model selection techniques of `forward selection` or `backward elimination` will be also quite interesting as more efficient models could be found out. 

### Organization

- Provide constructive feedback on any issues with file and/or code organization.

**Modularization**: Break down the code into separate modules or functions. For example, you can have separate functions for data loading, preprocessing, visualization, and modeling. This makes the code more organized and reusable. Additionlly, if you create a function for the modelling part, it would be easier to apply model selection techniques, as you can loop through different combinations of features.

### Further comments

- (Optional) Any further comments or feedback?
- 
In general the project is quite interesting and has a lot of potential to explore. You could also benefit a lot from using other models like Random Forest, XGboost or LightGBM, which are quite powerful when in comes to regression prediction, and are proven to be more accurate than linear regression.
